From 028e378f453cd9ca6eeb0055ebb5b7b3b793b364 Mon Sep 17 00:00:00 2001
From: "Tsuchiya Yuto (kitakar5525)" <kitakar@gmail.com>
Date: Mon, 20 Apr 2020 22:15:15 +0900
Subject: [PATCH 3/5] mwifiex_pcie: disable parent bridge_d3

On Windows, it seems that wifi parent bridge will not enter D3 state
(stay on D0 state) [1]. And on Linux, disabling the D3 state for the
bridge fixes driver crashing after suspend.

This commit disables the parent bridge D3 state on driver initialization
to fix driver crashing after suspend.

NOTE_1: The driver does not crash on LTS 4.4 series but does crash on
        LTS 4.9 series. Maybe commit 9d26d3a8f1b0c4 ("PCI: Put PCIe ports
        into D3 during suspend") changed the bridge D-state behavior.

NOTE_2: As a side effect, it reports parent bridge is in D3hot. I'm not
        sure why.

        acpi device:4b: Cannot transition to power state D0 for parent in D3hot

        Power state of the parent bridge on SB1 before this commit:

        $ grep -H . /sys/bus/pci/devices/0000:00:1d.3/firmware_node/{power_state,real_power_state}
        /sys/bus/pci/devices/0000:00:1d.3/firmware_node/power_state:D0
        /sys/bus/pci/devices/0000:00:1d.3/firmware_node/real_power_state:D3hot

        Power state of the parent bridge on SB1 after this commit:

        $ grep -H . /sys/bus/pci/devices/0000:00:1d.3/firmware_node/{power_state,real_power_state}
        /sys/bus/pci/devices/0000:00:1d.3/firmware_node/power_state:D3hot
        /sys/bus/pci/devices/0000:00:1d.3/firmware_node/real_power_state:D3hot

NOTE_3: This change is needed for SP4 and later generations that use
        mwifiex (i.e. SP4/SP5/SP6, SB1/SB2 and SL1/SL2). Not needed for
        Surface 3 but safe to apply as this change doesn't break S0ix.
        Not sure if needed for SP1/SP2 and SP3. Anyway, I think applying
        this change to all the devices may be safe regarding S0ix because
        Surface 3 can achieve S0ix anyway even when this patch applied.
        So, let's apply this change for all the devices for now.

[1] https://github.com/jakeday/linux-surface/issues/554#issuecomment-525262259
Signed-off-by: Tsuchiya Yuto (kitakar5525) <kitakar@gmail.com>
---
 drivers/net/wireless/marvell/mwifiex/pcie.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/net/wireless/marvell/mwifiex/pcie.c b/drivers/net/wireless/marvell/mwifiex/pcie.c
index 8722de6c7ceaa..b51c5e3571426 100644
--- a/drivers/net/wireless/marvell/mwifiex/pcie.c
+++ b/drivers/net/wireless/marvell/mwifiex/pcie.c
@@ -240,8 +240,13 @@ static int mwifiex_pcie_probe(struct pci_dev *pdev,
 					const struct pci_device_id *ent)
 {
 	struct pcie_service_card *card;
+	struct pci_dev *parent_pdev = pci_upstream_bridge(pdev);
 	int ret;
 
+	/* disable bridge_d3 to fix driver crashing after suspend on gen4+
+	 * Surface devices */
+	parent_pdev->bridge_d3 = false;
+
 	pr_debug("info: vendor=0x%4.04X device=0x%4.04X rev=%d\n",
 		 pdev->vendor, pdev->device, pdev->revision);
 
-- 
2.26.2

